### PR TITLE
[NJT] support chunk on batch dim

### PR DESCRIPTION
**Pull request by yaochengji**
- support chunk op on batch dim
- support empty_like op
- add tests for the like ops

cc @YuqingJ @jbschlosser @soulitzer

### CODE REVIEW

File: torch/nested/_internal/ops.py [hunk 1]
```diff
@@ -547,12 +551,43 @@ def chunk_default(func, *args, **kwargs):
...
+        dim0_size = inp._size[0]
+        chunk_size = math.ceil(dim0_size / chunks)
+
+        # get _offsets of the chunks
+        chunk_offsets = [
+            inp._offsets[i * chunk_size + 1 : (i + 1) * chunk_size + 1]
+            - inp._offsets[i * chunk_size]
+            for i in range(chunks)
+        ]
+        chunk_offsets = [F.pad(x, (1, 0), value=0) for x in chunk_offsets]
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:569]: Might be simpler if you converted to lengths, because then you can just apply chunk on the lengths as-is instead having to derive chunk_size yourself.
```
>>> offsets = torch.tensor([0, 4, 5, 9, 17])
>>> lengths = offsets.diff()
>>> lengths
tensor([4, 1, 4, 8])
>>> chunked_lengths = lengths.chunk(3)
>>> chunked_lengths
(tensor([4, 1]), tensor([4, 8]))
>>> torch.cat([torch.tensor([0]), torch.cumsum(chunked_lengths[0], dim=0)])
tensor([0, 4, 5])
```

**Review by soulitzer — COMMENTED** [review]
Thanks for the contribution! I added some comments on how chunk might be computed differently.

(Also, small logistical note is that we're calling these tensors "NJT"s now, so updating the title to reflect that)

File: torch/nested/_internal/ops.py (hunk 1 above)
Comment (yaochengji) [reply torch/nested/_internal/ops.py:569]: @soulitzer thanks for reviewing. Could you also explain why do you use `torch.cat` here? I used `F.pad` instead of `torch.cat` because it could avoid creating on demand tensor. (torch.cat would create the on demand tensor `torch.tensor([0])`

Comment (soulitzer) [reply torch/nested/_internal/ops.py:569]: You're right, we do create an extra tensor in this case. F.pad sounds good to me.

**Comment by yaochengji** [comment]
@soulitzer thanks, your comments have already been addressed.

File: torch/nested/_internal/ops.py [hunk 2]
```diff
@@ -21,11 +22,13 @@ def _outer_to_inner_dim(ndim, dim):
...
 
-def _wrap_jagged_dim(ndim, dim, op_name, convert_to_inner_dim=True):
+def _wrap_jagged_dim(
+    ndim, dim, op_name, convert_to_inner_dim=True, allow_dim_zero=False
+):
     from torch._prims_common import canonicalize_dims
 
     wrapped = canonicalize_dims(ndim, dim)
-    if wrapped < 2:
+    if wrapped == 1 or (wrapped == 0 and not allow_dim_zero):
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:31]: nit: if we're going to do this, we should make sure the error message no longer mentions dim=0 since that's supported now

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -21,11 +22,13 @@ def _outer_to_inner_dim(ndim, dim):
     return 0 if dim < 2 else dim - 1
 
 
-def _wrap_jagged_dim(ndim, dim, op_name, convert_to_inner_dim=True):
+def _wrap_jagged_dim(
+    ndim, dim, op_name, convert_to_inner_dim=True, allow_dim_zero=False
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:26]: tiny nit: `allow_batch_dim` as a more conceptual name

File: torch/nested/_internal/ops.py [hunk 4]
```diff
@@ -547,12 +551,38 @@ def chunk_default(func, *args, **kwargs):
...
+    if new_kwargs["dim"] == 0:
+        chunks = new_kwargs["chunks"]
+        dim0_size = inp._size[0]
+        chunk_size = math.ceil(dim0_size / chunks)
+
+        # get _offsets of the chunks
+        lengths = inp._offsets.diff()
+        chunked_lengths = lengths.chunk(chunks)
+        chunked_offsets = [torch.cumsum(x, dim=0) for x in chunked_lengths]
+        chunked_offsets = [F.pad(x, (1, 0), value=0) for x in chunked_offsets]
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:567]: nit: it's not more elegant, but offsets are constructed for the typical NJT construction case here:
https://github.com/pytorch/pytorch/blob/86dedebeafdd7b08d21432cebd7538437d3b7509/torch/nested/_internal/nested_tensor.py#L339-L344

I see that @soulitzer already pointed this out :) I think we should factor out a nice helper function with the faster impl to avoid any redundancy

**Review by jbschlosser — COMMENTED** [review]
couple nits. Thanks for the PR!

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (yaochengji) [reply torch/nested/_internal/ops.py:567]: @jbschlosser , thanks for reviewing. The two methods for constructing offsets differ not in elegance but in compatibility with other code. The F.pad approach offers better compatibility, as it avoids creating leaf tensors on demand.

Using DTensor as an example, the `distribute_module` function only converts tensors created in the `__init__` function to DTensor, as well as the inputs and outputs of the module. It does not handle on-demand creation of tensors, here it is `torch.zeros(1, dtype=torch.int64, device=values.device)` during the forward. These on-demand tensors are ordinary tensors, which can cause sharding propagation to fail.

**Comment by yaochengji** [comment]
@jbschlosser I addressed your error message and `allow_batch_dim` comments, and add `xfailIfTorchDynamo` to the like_xxx ops test

**Comment by yaochengji** [comment]
@soulitzer @jbschlosser  any further thoughts about the `torch.cat` or `F.pad` method? Or is there any other comment about this PR?

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:571]: Thanks for the context! Given the compatibility issues you pointed out, it may make sense to change the typical NTJ construction path to match the padded impl here. For now at least, can you make a note that the two impls differ with a TODO for synchronizing them?

**Comment by yaochengji** [comment]
@jbschlosser thanks, a TODO is added

**Review by jbschlosser — APPROVED** [review]

**Comment by yaochengji** [comment]
@jbschlosser @YuqingJ the CI test should be able to pass, could you merge again?
